flink-clients/src/main/java/org/apache/flink/client/program/StreamContextEnvironment.java

```diff
@@ -303,6 +303,7 @@ private void checkMainConfiguration(Configuration clusterConfigMap, List<String>
     private void checkCheckpointConfig(Configuration clusterConfigMap, List<String> errors) {
         CheckpointConfig expectedCheckpointConfig = new CheckpointConfig();
         expectedCheckpointConfig.configure(clusterConfigMap);
+        configureCheckpointStorage(clusterConfigMap, expectedCheckpointConfig);
```

Comment:
The created `CheckpointStorage` may not be equal to the actual `CheckpointStorage` if the `equals()` method is not implemented, in which case it will compare if they are the same java instance.
How about to avoid doing this, but instead set `filesystem` to `state.checkpoint-storage` when `CheckpointConfig#setCheckpointStorage(...)` is invoked.

Reply:
Using state.checkpoint-storage for comparison is not entirely accurate, as it only checks if the Storage class name is the same. The root cause of this issue is the lack of robustness in the current mechanism for checking the checkpoint storage. Therefore, I prefer not to address this issue in this PR. Additionally, we are planning to deprecate the storage setter interface soon and migrate all configuration options into the configuration. I would prefer to update the checkpoint storage checking mechanism after the migration is completed.